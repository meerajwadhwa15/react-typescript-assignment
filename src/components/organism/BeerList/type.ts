export type listType = {
  id: number
  image_url: string
  name: string
  tagline: string
  volume: {
    value: number
    unit: string
  }
}

export interface ComponentProps {}

export interface ComponentState {}

export interface ReducerState {
  list: listType[]
  isFetching: boolean
  errorMessage: string
}

export interface DispatchProps {
  fetchBeersAction: () => void
}

export interface BeerListReducerType {
  BeerListReducer: {
    list: listType[]
    isFetching: boolean
    errorMessage: string
  }
}

export interface ReducerAction {
  type: string
  list: listType[]
  isFetching: boolean
  message: string
}

export type FetchSuccessActionType = {
  type: string
  list: listType[]
}

export interface SetFetch {
  type: string
  isFetching: boolean
}

export type FetchErrorActionType = {
  type: string
  message: string
}
