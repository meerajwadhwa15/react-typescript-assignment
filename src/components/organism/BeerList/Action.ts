import { AnyAction } from 'redux'
import { AxiosResponse, AxiosError } from 'axios'
import { FetchSuccessActionType, SetFetch, FetchErrorActionType } from './type'
import { ThunkDispatch } from 'redux-thunk'
import Fetch, { ServerProps } from './../../../libs/fetch'
import { ERROR_MESSAGE } from './../../../config/message'
import { BEER_LIST } from './../../../config/api'

export const isFetching = (isFetching: boolean): SetFetch => {
  return { type: 'SET_FETCHING', isFetching }
}

export const FetchBeers = () => {
  return async (dispatch: ThunkDispatch<{}, {}, AnyAction>): Promise<void> => {
    const showLoader: SetFetch = isFetching(true)
    dispatch(showLoader)
    const serverProps: ServerProps = {
      url: BEER_LIST,
    }

    return Fetch(serverProps)
      .then((response: AxiosResponse) => {
        const fetchBeersSuccess: FetchSuccessActionType = {
          type: 'FETCH_BEERS_SUCCESS',
          list: response.data,
        }
        dispatch(fetchBeersSuccess)
      })
      .catch((err: AxiosError) => {
        const fetchBeersSuccess: FetchErrorActionType = {
          type: 'FETCH_BEERS_ERROR',
          message: ERROR_MESSAGE,
        }
        dispatch(fetchBeersSuccess)
      })
      .finally(() => {
        const hideLoader: SetFetch = isFetching(false)
        dispatch(hideLoader)
      })
  }
}
