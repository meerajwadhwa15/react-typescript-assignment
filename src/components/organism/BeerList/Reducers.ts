import { Reducer } from 'redux'
import { ReducerState, ReducerAction } from './type'

export const initialState: ReducerState = {
  list: [],
  isFetching: false,
  errorMessage: '',
}

const BeerListReducer: Reducer<ReducerState, ReducerAction> = (
  state = initialState,
  action
) => {
  switch (action.type) {
    case 'SET_FETCHING':
      return { ...state, isFetching: action.isFetching }
    case 'FETCH_BEERS_SUCCESS':
      return { ...state, errorMessage: '', list: action.list }
    case 'FETCH_BEERS_ERROR':
      return { ...state, errorMessage: action.message }
    default:
      return state
  }
}

export default BeerListReducer
