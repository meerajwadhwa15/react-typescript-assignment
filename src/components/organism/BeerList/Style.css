.cards {
  width: 640px;
}