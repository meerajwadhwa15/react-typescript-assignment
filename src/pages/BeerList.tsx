import React, { FunctionComponent } from 'react'
import Layout from './../components/templates/Layout'
import BeerList from './../components/organism/BeerList'

type BeerListProps = {}

const BeerListPage: FunctionComponent<BeerListProps> = (props) => (
  <Layout>
    <BeerList />
  </Layout>
)

export default BeerListPage
