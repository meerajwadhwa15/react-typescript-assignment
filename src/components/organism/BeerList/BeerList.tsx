import React, { ReactNode, Component, Fragment } from 'react'
import { connect } from 'react-redux'
import { ThunkDispatch } from 'redux-thunk'
import ListCard from './../../molecules/ListCard'
import {
  listType,
  ComponentState,
  BeerListReducerType,
  ReducerState,
  DispatchProps,
  ComponentProps,
} from './type'
import { FetchBeers } from './Action'
import Loader from './../../molecules/Loader'
import ErrorMessage from './../../atoms/ErrorMessage'
import NoRecordFound from './../../atoms/NoRecordFound'
import './Style.css'

export const mapStateToProps = ({
  BeerListReducer: { list, isFetching, errorMessage },
}: BeerListReducerType): ReducerState => ({
  list,
  isFetching,
  errorMessage,
})

export const mapDispatchToProps = (
  dispatch: ThunkDispatch<{}, {}, any>
): DispatchProps => ({
  fetchBeersAction: () => {
    dispatch(FetchBeers())
  },
})

export type Props = ReducerState & DispatchProps

export class BeerList extends Component<Props, ComponentState> {
  static defaultProps = {
    list: [],
  }

  componentDidMount() {
    const { list } = this.props
    if (!list.length) {
      this.props.fetchBeersAction()
    }
  }

  render(): ReactNode {
    const { list, isFetching, errorMessage } = this.props

    return (
      <div className="cards">
        {isFetching ? (
          <Loader />
        ) : errorMessage ? (
          <ErrorMessage>{errorMessage}</ErrorMessage>
        ) : list.length ? (
          list.map((item: listType) => <ListCard key={item.id} item={item} />)
        ) : (
          <NoRecordFound>No Record Found.</NoRecordFound>
        )}
      </div>
    )
  }
}

// Now let's connect our component!
export default connect<
  ReducerState,
  DispatchProps,
  ComponentProps,
  BeerListReducerType
>(
  mapStateToProps,
  mapDispatchToProps
)(BeerList)
