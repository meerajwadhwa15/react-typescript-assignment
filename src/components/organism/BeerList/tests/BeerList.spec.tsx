import React from 'react'
import { shallow } from 'enzyme'
import { BeerList, mapStateToProps, Props } from './../BeerList'
import { create, ReactTestInstance } from 'react-test-renderer'
import { BrowserRouter } from 'react-router-dom'

const sampleLoadingMockData: Props = {
  fetchBeersAction: jest.fn(),
  list: [],
  isFetching: true,
}

const sampleListMockData: Props = {
  fetchBeersAction: jest.fn(),
  list: [
    {
      id: 1,
      image_url: 'https://images.punkapi.com/v2/keg.png',
      name: 'Buzz',
      tagline: 'A Real Bitter Experience.',
      volume: {
        value: 20,
        unit: 'litres',
      },
    },
    {
      id: 2,
      image_url: 'https://images.punkapi.com/v2/keg.png',
      name: 'Buzz 2',
      tagline: 'A Real Bitter Experience.',
      volume: {
        value: 20,
        unit: 'litres',
      },
    },
  ],
  isFetching: false,
}

const ComponentWithLoading = () => (
  <BrowserRouter>
    <BeerList {...sampleListMockData} />
  </BrowserRouter>
)

const ComponentWithList = () => (
  <BrowserRouter>
    <BeerList {...sampleLoadingMockData} />
  </BrowserRouter>
)

describe('BeerList Enzyme testing', () => {
  it('renders with Loading without crashing', () => {
    shallow(<ComponentWithLoading />)
  })

  it('renders with beerList without crashing', () => {
    shallow(<ComponentWithList />)
  })

  test('Matches the snapshot with beer List', () => {
    const component = create(<ComponentWithList />)
    expect(component.toJSON()).toMatchSnapshot()
  })

  test('Matches the snapshot without beer list', () => {
    const component = create(<ComponentWithLoading />)
    expect(component.toJSON()).toMatchSnapshot()
  })

  test('Should call fetchBeersAction ', () => {
    const spy = jest.spyOn(sampleLoadingMockData, 'fetchBeersAction')
    const component = create(<BeerList {...sampleLoadingMockData} />)
    const instance: ReactTestInstance = component.getInstance()
    instance.componentDidMount()
    expect(spy).toBeCalled()
  })

  test('Should test mapStateToProps', () => {
    expect(
      mapStateToProps({
        BeerListReducer: {
          list: [],
          isFetching: false,
        },
      })
    ).toEqual({
      list: [],
      isFetching: false,
    })
  })
})
