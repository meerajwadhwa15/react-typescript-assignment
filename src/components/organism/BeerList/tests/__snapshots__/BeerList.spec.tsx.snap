// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BeerList Enzyme testing Matches the snapshot with beer List 1`] = `
<div
  className="cards"
>
  <div
    className="loader"
  >
    Loading...
  </div>
</div>
`;

exports[`BeerList Enzyme testing Matches the snapshot without beer list 1`] = `
<div
  className="cards"
>
  <div
    className="card"
  >
    <a
      href="/1"
      onClick={[Function]}
      title="Buzz"
    >
      <div>
        <img
          alt="A Real Bitter Experience."
          height="250"
          src="https://images.punkapi.com/v2/keg.png"
          width="150"
        />
      </div>
      <div>
        <h2>
          Buzz
        </h2>
        <div>
          A Real Bitter Experience.
        </div>
        <div>
          20
           
          litres
        </div>
      </div>
    </a>
  </div>
  <div
    className="card"
  >
    <a
      href="/2"
      onClick={[Function]}
      title="Buzz 2"
    >
      <div>
        <img
          alt="A Real Bitter Experience."
          height="250"
          src="https://images.punkapi.com/v2/keg.png"
          width="150"
        />
      </div>
      <div>
        <h2>
          Buzz 2
        </h2>
        <div>
          A Real Bitter Experience.
        </div>
        <div>
          20
           
          litres
        </div>
      </div>
    </a>
  </div>
</div>
`;
